common/djangoapps/entitlements/signals.py

```diff
@@ -0,0 +1,16 @@
+"""
+Entitlements related signal handlers.
+"""
+
+from django.dispatch import receiver
+
+from entitlements.models import CourseEntitlement
+from student.signals import UNENROLL_DONE
+
+
+@receiver(UNENROLL_DONE)
+def unenroll_entitlement(sender, course_enrollment=None, skip_refund=False, **kwargs):  # pylint: disable=unused-argument
+    """
+    Un-enroll user from entitlement upon course run un-enrollment if exist.
+    """
+    CourseEntitlement.unenroll_entitlement(course_enrollment)
```

Comment:
I am concerned that we need to look at the code for other locations that we are already unenrolling based on an entitlement.

For example there are a few cases in the view code here: https://github.com/edx/edx-platform/blob/c6ec5edbf76d5b5212872e5bbbe6adde8fe6ed9b/common/djangoapps/entitlements/api/v1/views.py#L40-L40

That we handle both cases.  We might need to refactor that code so that this signal is not called multiple times resulting in unnecessary DB interactions.

Reply:
@staubina I don't think there will be any duplicate call of that signal, if unenrollment call is from entitlement api then signal will be fired only once from inside of this unenroll call. https://github.com/edx/edx-platform/blob/c6ec5edbf76d5b5212872e5bbbe6adde8fe6ed9b/common/djangoapps/entitlements/api/v1/views.py#L44